### ISSUE TITLE
VSCode plugin on save garbles up the file.

### ISSUE THREAD

**Issue by mctrafik** [biomejs/biome#2592]
### Environment information

```block
CLI:
  Version:                      1.6.3
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v20.11.1"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "npm/10.2.4"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           true
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 false

Workspace:
  Open Documents:               0
```

### What happened?

We use prettier formatter because it supports plugins that biome doesn't.

We perform formatting on save.
We also run quickFix by biome on save as well.

VSCOde settings:

```
 "editor.defaultFormatter": "esbenp.prettier-vscode",
 "editor.formatOnSave": true,
 "editor.codeActionsOnSave": {
    "quickfix.biome": "explicit",
    "source.organizeImports.biome": "explicit",
```

But recently we had to turn both of the biome features off because when any biome quickfix seems to randomly remove *LARGE CHUNKS OF THE FILE*. Sometimes the corruption is just a line or two. Sometimes it removes the last 70% of the file. 

### Expected result

What I would expect is for biome to not garble up the file! 

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
Are you able to provide a reproduction?

**Comment by zauni** [comment]
VSCode unfortunately has a bad history with Prettier formatting and code actions after save (https://github.com/prettier/prettier-vscode/issues/716) It looks a little bit as the prettier format is too slow and the codeActionsOnSave are triggered before prettier is finished. I don't know if there is a "codeActionsAfterFormat" or something (e.g. Zed editor is doing it like that)

**Comment by mctrafik** [comment]
@zauni I see. So it's an interaction where `prettier` writes later, and `biome` does a write before that and that's whey when I added biome I started seeing issues? 

Well, for me personally, I'll keep prettier on save, and will turn off biome on save. The reason being formatting pretty much always a safe operation. Where as some of the biome rules I have enabled make development annoying, i.e. if I copy an import from another file and then start writing code to use said import, if I pause (which triggers auto save), the import I just copied gets removed. 

Thanks for the explanation!

P.S. When biome formatting has more features, will gladly switch to it from prettier.
